### ISSUE TITLE
Module suggestion: CAPTCHA & reCAPTCHA

### ISSUE THREAD

**Issue by rooby** [govCMS/GovCMS7#671]
Since webforms are used on GovCMS there really needs to be some sort of CAPTCHA to prevent spamming. 

I would say the reCAPTCHA "I'm not a robot" checkbox is probably the best one I've used and also is familiar to users.

**Comment by tobybellwood** [comment]
S'pose the wider question is, do webform receivers get a lot of spam?

We've got Honeypot in place across govCMS, and we don't see much bot spam ourselves through our webforms.  Are other agencies experiences different?

**Comment by rooby** [comment]
In my experience you don't until you do, and then you get it good.

But also, I think user registrations are generally more of a target than webforms.

You can also implement reCAPTCHA in your theme relatively easily.

**Comment by Feng-Shui** [comment]
Accessibility of captchas can be pretty poor, Would be good to see anything going into the profile get some testing on this front.  We've used Honeypot successfully with webform.

**Comment by rooby** [comment]
Actually it's not super trivial to manually implement recaptcha in the theme.

**Comment by rooby** [comment]
Honeypot is not bad but it doesn't always stop bots.

Google has this page on the accessibility of ReCAPTCHA. It's quite accessible as far as CAPTCHAs go.
https://support.google.com/recaptcha/answer/6175971?hl=en

**Comment by rooby** [comment]
Remember this is not to force all sites to use reCAPTCHA, just to give it as an option for site builders since it is the go-to CAPTCHA solution.

**Comment by rooby** [comment]
One potential issue is the use of 3rd party javascripts, but that is already the case with Google Analytics so I wouldn't expect it to be an issue here.

**Comment by WebProject2015** [comment]
We had some heavy spamming of one of our web forms last week, including it finding a way through the honeypot module, so anything that makes it easy for us to put captcha on high profile web forms would be something we'd be interested in. Over the course of 48 hours, we have some 5800 web form submissions through one form. At this stage we're not sure what fixed it - whether it was the blocking of the originating IP addresses and changes made to the honeypot module so anything that prevents a reoccurrence would be great. We'll look at potentially do it within the theme but a module would be preferred.

**Comment by pandaskii** [comment]
Regarding honeypot module, it is basically a honeypot spam prevention method alongside a time-based protection (users can't submit a form in less than X seconds, and X increases exponentially with each consecutively failed submission).

It is not a cure-all medicine but quite an effective method fighting with automated spambots.

In a real site, we may have to use honeypot with other methods together since the bots evolve rapidly and become smarter.

With few simple and basic methods can be used:
- Set the field name in a natural way, for example, "first_name" is better than "faq". BTW, the default value is "URL" so does the spambot know :)
- Change the field name in certain time period especially when the site is under heavy spamming.
- Maybe change your webform URL as well. For example, from "/contact" to "/contact-us" 

Regarding Captcha/ReCaptcha, yes, many people dislike it because it is a method that "punishing everyone for one person's mistake" etc.

My two cents, this is a business decision for each site. If theme layer can do it and the usage is few, why bother a module?

**Comment by rooby** [comment]
It's not trivial to implement recaptcha in the theme layer. Even if not many sites use webforms I think it needs to be there for those that do.

Depending on the form you cannot set the time restriction too high or you will affect end users way more than recaptcha would (at least last time I checked, the user experience is terrible when you get that honeypot time error). Especially for forms where the users may be likely to use browser autofill.

I can confirm on a live SaaS site, that honeypot alone is insufficient. The scale of spam submissions remains to be seen though. If it's a negligible amount then that would likely fall to site admins to filter out for themselves. After all, nothing will stop manually created spam from a determined person.

**Comment by fubarhouse** [comment]
I just wanted to add that I've added some really effective validators to a form at work with a huge success rate.

If there was a way of utilising these - even as a govcms module for custom validators where somebody could simply 'add' the validator set to an existing form, I would be prepared to share some details.

If it can't be systematically implemented (even by opt-in), there's little point providing it as a feature but simply as a independently handled task.

We opted to not use reCAPTCHA because it is interactive - non-interactive which isn't yet supported would be wonderful.
